livekit/src/room/token_source/mod.rs

```diff
@@ -0,0 +1,304 @@
+    fn fetch_synchronous(&self, options: &TokenSourceFetchOptions) -> Result<TokenSourceResponse, Box<dyn Error>> {
+        let MinterCredentials { url: server_url, api_key, api_secret } = self.credentials_source.get();
+
+        // FIXME: apply options in the below code!
+        let participant_token = access_token::AccessToken::with_api_key(&api_key, &api_secret)
+            .with_identity("rust-bot")
+            .with_name("Rust Bot")
+            .with_grants(access_token::VideoGrants {
+                room_join: true,
+                room: "my-room".to_string(),
+                ..Default::default()
+            })
+            .to_jwt()?;
+
+        Ok(TokenSourceResponse { server_url, participant_token })
+    }
+}
+
+impl Default for TokenSourceMinter<MinterCredentialsEnvironment> {
+    fn default() -> Self {
+        Self::new(MinterCredentialsEnvironment::default())
+    }
+}
+
+
+
+
+
+
+pub struct TokenSourceCustomMinter<
+    MintFn: Fn(access_token::AccessToken) -> Result<String, AccessTokenError>,
+    Credentials: MinterCredentialsSource,
+> {
+    mint_fn: MintFn,
+    credentials_source: Credentials,
+}
+
+impl<
+    MF: Fn(access_token::AccessToken) -> Result<String, AccessTokenError>,
+    CS: MinterCredentialsSource
+> TokenSourceCustomMinter<MF, CS> {
+    pub fn new_with_credentials(mint_fn: MF, credentials_source: CS) -> Self {
+        Self { mint_fn, credentials_source }
+    }
+    pub fn new(mint_fn: MF) -> TokenSourceCustomMinter<MF, MinterCredentialsEnvironment> {
+        TokenSourceCustomMinter::new_with_credentials(mint_fn, MinterCredentialsEnvironment::default())
+    }
+}
+
+
+impl<
+    MF: Fn(access_token::AccessToken) -> Result<String, AccessTokenError>,
+    C: MinterCredentialsSource,
+> TokenSourceFixedSynchronous for TokenSourceCustomMinter<MF, C> {
+    fn fetch_synchronous(&self) -> Result<TokenSourceResponse, Box<dyn Error>> {
+        let MinterCredentials { url: server_url, api_key, api_secret } = self.credentials_source.get();
+
+        let participant_token = (self.mint_fn)(access_token::AccessToken::with_api_key(&api_key, &api_secret))?;
+
+        Ok(TokenSourceResponse { server_url, participant_token })
+    }
+}
+
+
+
+
+
+use reqwest::{header::HeaderMap, Method};
+
+pub struct TokenSourceEndpoint {
+    url: String,
+    method: Method,
+    headers: HeaderMap,
+}
+
+impl TokenSourceEndpoint {
+    pub fn new(url: &str) -> Self {
+        Self {
+            url: url.into(),
+            method: Method::POST,
+            headers: HeaderMap::new(),
+        }
+    }
+}
+
+impl TokenSourceConfigurable for TokenSourceEndpoint {
+    async fn fetch(&self, options: &TokenSourceFetchOptions) -> Result<TokenSourceResponse, Box<dyn Error>> {
+        let client = reqwest::Client::new();
+
+        // FIXME: What are the best practices around implementing Into on a reference to avoid the
+        // clone?
+        let request_body: TokenSourceRequest = options.clone().into();
```

Comment:
thought:
- You currently implement `impl Into<TokenSourceRequest> for TokenSourceFetchOptions`
- Try implementing `impl Into<TokenSourceRequest> for &TokenSourceFetchOptions` instead